### PR TITLE
Generalize Geojson to VOC Annotation Tool

### PR DESCRIPTION
**Pull request by conor-wallace**
Generalize the ```geo_convert.py``` script from ```ml-solar-outline-detection``` to be used in any case where detection data is given in geojson polygons that need to be easily converted to VOC and Rotated VOC annotations.

### CODE REVIEW

File: ml_dronebase_data_utils/__init__.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-from . import pascal_voc, s3  # noqa: F401
+from . import convert_geojson, pascal_voc, s3, visualize  # noqa: F401
 
 __author__ = "Conor Wallace"
-__version__ = "0.0.2"
+__version__ = "0.0.6"
```
Comment (conor-wallace) [inline ml_dronebase_data_utils/__init__.py:4]: This was a discrepancy from moving the public package to private, so now I'm pushing the version past the public version for continuity.

File: ml_dronebase_data_utils/box_utils.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        top_right = pxl_points_sorted[1]
+        bottom_right = pxl_points_sorted[2]
+        bottom_left = pxl_points_sorted[3]
+
+        logging.debug(f"top left = {top_left}")
+        logging.debug(f"top right = {top_right}")
+        logging.debug(f"bottom right = {bottom_right}")
+        logging.debug(f"bottom left = {bottom_left}")
+
+        vec1_mag = np.linalg.norm(bottom_right - top_right)
```
Comment (conor-wallace) [inline ml_dronebase_data_utils/box_utils.py:49]: The rotation angle is measure w.r.t. the width vector (e.g., the smaller side) of the box.

File: ml_dronebase_data_utils/box_utils.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        y1 (float): y coordinate of rotation point
+        x2 (float): x coordinate of reference point
+        y2 (float): y coordinate of reference point
+
+    Returns:
+        float: orientation angle
+    """
+    angle = np.degrees(np.arctan2(y2 - y1, x2 - x1))
+    logging.debug(f"Before angle = {angle}")
+    if (-90 <= angle) and (angle <= 90):
```
Comment (conor-wallace) [inline ml_dronebase_data_utils/box_utils.py:113]: The rotation angle is constrained to the range -pi/4 to pi/4, so if it is outside this range, convert it.

Comment (conor-wallace) [reply ml_dronebase_data_utils/box_utils.py:113]: This will help reduce the number of anchor angles from the suggested -pi/2 to pi/2 when training a RRPN.

File: ml_dronebase_data_utils/box_utils.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    boxes = boxes[sorted_idxs]
+
+    vertices = []
+    for i in range(num_instances):
+        v = extract_vertices(boxes[i])
+        vertices.append(v)
+    return np.asarray(vertices)
+
+
+def rotated_boxes_to_vertices(
```
Comment (conor-wallace) [inline ml_dronebase_data_utils/box_utils.py:141]: Borrowed from ```detectron2.utils.visualize.draw_rotated_box_with_label```

File: ml_dronebase_data_utils/box_utils.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    return np.asarray(vertices)
+
+
+def extract_vertices(box: np.ndarray) -> List[List[float]]:
+    xmin, ymin, xmax, ymax = box
+    vertices = np.array([[xmin, ymin], [xmax, ymin], [xmax, ymax], [xmin, ymax]])
+    return vertices
+
+
+def extract_rotated_vertices(
```
Comment (conor-wallace) [inline ml_dronebase_data_utils/box_utils.py:169]: Adapted from ```detectron2.utils.visualize.draw_rotated_box_with_label```. The ```box_mode``` arg allows you to use this function outside of detectron2 since they have a strict requirement that rotated boxes are represented with the ```XYWHA_ABS``` box mode, whereas we use the default VOC definition of ```XYXYA_ABS```.

File: ml_dronebase_data_utils/box_utils.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    theta = angle * math.pi / 180.0
+    c = math.cos(theta)
+    s = math.sin(theta)
+    deltas = [(-w / 2, h / 2), (-w / 2, -h / 2), (w / 2, -h / 2), (w / 2, h / 2)]
+
+    vertices = []
+    for k in range(4):
+        x_delta = deltas[k][0]
+        y_delta = deltas[k][1]
+        x = x_delta * c - y_delta * s + xc
```
Comment (conor-wallace) [inline ml_dronebase_data_utils/box_utils.py:196]: This differs from detectron2's implementation of this function likeso:

```python
theta = angle * math.pi / 180.0
c = math.cos(theta)
s = math.sin(theta)
rect = [(-w / 2, h / 2), (-w / 2, -h / 2), (w / 2, -h / 2), (w / 2, h / 2)]
rotated_rect = [(s * yy + c * xx + cnt_x, c * yy - s * xx + cnt_y) for (xx, yy) in rect]
```

**Comment by conor-wallace** [comment]
Additional information on the approach to computing rotation angles can be seen in [this](https://dronebase.atlassian.net/wiki/spaces/AIML/pages/106528947/Rotated+Bounding+Boxes) document on confluence.

**Review by isaaccorley — APPROVED** [review]
Lgtm, thanks for adding tests and typing to make it readable.

File: tests/test_convert_geojson.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    assert np.round(boxes[0][-1], decimals=2) == 63.43
+
+
+def test_get_boxes(ortho_path, geo_path):
+    ortho = rasterio.open(ortho_path)
+    gdf = gpd.read_file(geo_path)
+
+    vertices = get_pixel_vertices(ortho, gdf)
+    boxes = vertices_to_boxes(vertices)
+    assert boxes is not None
```
Comment (jonlwowski012) [inline tests/test_convert_geojson.py:74]: Should we have a stricter test than just checking if it is None?

File: tests/test_convert_geojson.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def test_clockwise_vertices_to_rotated_boxes(vertices_clockwise):
+    boxes = vertices_to_rotated_boxes(vertices_clockwise)
+    assert boxes[0][-1] == -45.0
+
+
+def test_counter_clockwise_vertices_to_rotated_boxes(vertices_counter_clockwise):
+    boxes = vertices_to_rotated_boxes(vertices_counter_clockwise)
+    assert np.round(boxes[0][-1], decimals=2) == 63.43
```
[the remark this answers is not in the document]
Comment (conor-wallace) [reply tests/test_convert_geojson.py:65]: All four of the ```{}_vertices_to_rotated_boxes()``` tests were meant to test the edge cases of potential angle measurements and therefore were pre-computed so they are just a check that the code produces the expected result.

As for this particular case, it is a set of vertices that are rotated 63.43 degrees counterclockwise from the 0-degree box (a box with which the width vector is flat on the x-axis). If you calculate the angle of the width vector w.r.t. the positive x-axis, you would get 63.43 degrees, after rounding to the nearest hundredth. To see an example box similar to this test case, you can refer to the third figure with a rotation angle of 60 degrees in [this](https://dronebase.atlassian.net/wiki/spaces/AIML/pages/106528947/Rotated+Bounding+Boxes) document.

File: tests/test_convert_geojson.py (hunk 7 above)
Comment (conor-wallace) [reply tests/test_convert_geojson.py]: I've added ```shapely``` testing to show that the reconstructed vertices match the given geojson vertices.

File: tests/test_convert_geojson.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (conor-wallace) [reply tests/test_convert_geojson.py]: I removed this test as it was mostly for debugging and prototyping.
